src/auth/login.controller.ts

```diff
@@ -20,17 +36,81 @@ export class LoginController {
   })
   @Post('login')
   async login(@Req() req: Request, @Res() res: Response) {
-    const cookiesOptions: CookieOptions = {
-      httpOnly: true,
-      sameSite: 'none',
-      secure: true,
-    };
-    const { family, major } = useragent.parse(req.headers['user-agent']) || {};
-    if (family === 'Chrome' && +major >= 51 && +major <= 66) {
-      delete cookiesOptions.sameSite;
-      delete cookiesOptions.secure;
+    const { did, verifiedRoles } =
+      (req.user as { did: string; verifiedRoles: string[] }) || {};
+
+    if (!did) {
+      throw new UnauthorizedException();
     }
-    res.cookie('token', req.user, cookiesOptions);
-    return res.send({ token: req.user });
+
+    const cookiesOptions = this.cookiesServices.getCookiesOptionBasedOnUserAgent(
+      req.headers['user-agent'],
+    );
+
+    const [token, refreshToken] = await Promise.all([
+      this.tokenService.generateAccessToken({ did, verifiedRoles }),
+      this.tokenService.generateRefreshToken({
+        userDid: did,
+        verifiedRoles,
+      }),
+    ]);
+
+    res.cookie(
+      this.configService.get<string>('JWT_ACCESS_TOKEN_NAME'),
+      token,
+      cookiesOptions,
+    );
+
+    res.cookie(
+      this.configService.get<string>('JWT_REFRESH_TOKEN_NAME'),
+      refreshToken,
+      cookiesOptions,
+    );
+
+    return res.send({ token, refreshToken });
+  }
+
+  @Get('refresh_token')
+  async refreshToken(@Req() req: Request, @Res() res: Response) {
+    const refreshTokenString =
+      req.cookies[this.configService.get<string>('JWT_REFRESH_TOKEN_NAME')];
+
+    if (!refreshTokenString) {
+      throw new UnauthorizedException();
+    }
+
+    const { userDid, id, verifiedRoles } =
+      (await this.tokenService.verifyRefreshToken(refreshTokenString)) || {};
+
+    if (!id || !userDid) {
+      throw new UnauthorizedException();
+    }
+
+    const cookiesOptions = this.cookiesServices.getCookiesOptionBasedOnUserAgent(
+      req.headers['user-agent'],
+    );
+
+    const [token, refreshToken] = await Promise.all([
+      this.tokenService.generateAccessToken({ did: userDid, verifiedRoles }),
+      this.tokenService.generateRefreshToken({
+        userDid,
+        verifiedRoles,
```

Comment:
Should we be reusing the same `verifiedRoles` that we received from the old refresh token? I wonder if it might make more sense to refresh the roles so that the accessToken is up to date? If so, maybe we don't need to store the roles in the refreshToken

Reply:
👍